### DISCUSSION TITLE
Can I use arrow.core typed error classes in java?

### DISCUSSION THREAD

**Discussion by jeusdi** [arrow-kt/arrow#3397] [Q&A]
It's not clear to if I'm able to use arrow typed error classes in java.

I've took a look on [arrow typed errors documentation here](https://arrow-kt.io/learn/typed-errors/).

I mean, I'm aware I'm able to import maven artifact dependency into my java project, but it's not clear to how to use its classes. Am I able to use `Either`, or `EitherKt`.

Which limitations am I going to face with?

I've played a bit, but I'm getting some compilation warnings:

```java
	public static Either<NonEmptyList<String>, Reference> of(ReferenceUniqueIdentifier referenceUniqueIdentifier) {
		...
		return null;
	}
```
How could I provide an either object like `Either<NonEmptyList<String>, Reference`?

**Comment by serras** [comment]
I strongly recommend checking [Kotlin's docs](https://kotlinlang.org/docs/java-to-kotlin-interop.html) about calling Kotlin functions from Java. Whether the method is available through `Either` or `EitherKt` depends on whether it is a member or an extension method, respectively.

**Reply by nomisRev** [reply]
Doing this is extremely frustrating, this might be improved but we've never properly spend time since it would almost always also affects source -and binary compatibility.

Biggest annoyance is that due to `Nothing`, and the difference in how generics widen that you need to explicitly cast very often in Java. This makes is not very ideal, and if you cast incorrectly you still also end up with runtime exceptions.

**Reply by serras** [reply]
It seems that [vavr](https://docs.vavr.io/#_either) is the library of choice for working with `Either` in Java. Maybe we should improve our compatibility with it?

**Reply by nomisRev** [reply]
Yes, ideally we provide that out-of-the-box with some popular integrations but I think we need context parameters first.

```kotlin
context(_: Raise<String>)
fun io.vavr.Either<String, Int>.bind(): Int =
  fold({ raise(it) }) { it }
```

**Reply by serras** [reply]
But isn't the goal to make it easier for _Java_ developers to consume Arrow?

**Reply by nomisRev** [reply]
Yes, I was thinking of the opposite direction. We could provide a Java VAVR DSL that allows binding to Arrow's types, but I am unsure if the ergonomics of using Arrow overall would be okay. The variance with `Nothing` especially comes to mind.

A Java VAVR DSL could look like this (didn't check or test code):

```java
class Test {
  public static arrow.core.Either<String, Integer > one() {
      return (arrow.core.Either<String, Integer >) arrow.core.Either.Right(1);
  }
  
    public static io.vavr.Either<String, Integer> two() {
      return  io.vavr.Either.right(2);
  }
  
  public static io.vavr.Either<String, Integer> example() {
    return ArrowDSL.<String, Int>vavr((scope) -> {
      Integer one = scope.bind(one());
      Integer two = scope.bind(two());
      return one + two;
    });
  }
}
```
